File: programmers/불량 사용자.java
// Set 사용법 숙지. HashSet 또한 reference type이기 때문에 다른 Set에 넣을 때 new HashSet<>(set) 등과 같이 새로운 객채로 넣어야 한다.

import java.util.regex.Pattern;
import java.util.regex.Matcher;
import java.util.Set;
import java.util.HashSet;

class Solution {
    Set<Set<String>> set = new HashSet<>();

    public int solution(String[] user_id, String[] banned_id) {
        check(user_id, banned_id, 0, new boolean[user_id.length], new HashSet<String>());
        return set.size();
    }

    public void check (String[] user, String[] ban, int banIndex, boolean[] c, Set<String> list) {
        if (banIndex == ban.length) {
            set.add(new HashSet<>(list));
            return;
        }

        String regex = ban[banIndex].replace("*", ".");
        Pattern pattern = Pattern.compile(regex);

        for (int i = 0; i < user.length; i++) {
            if (!c[i]) {
                Matcher matcher = pattern.matcher(user[i]);
                if (matcher.matches()) {
                    c[i] = true;
                    list.add(user[i]);
                    check(user, ban, banIndex+1, c, list);
                    list.remove(user[i]);
                    c[i] = false;
                }
            }
        }
    }
}

File: programmers/불량 사용자.swift
// 정규표현식 연습

import Foundation

func solution(_ user_id:[String], _ banned_id:[String]) -> Int {
    var answerSet = Set<[String]>()

    let banIndexQueue = Queue<Int>()
    let checkQueue = Queue<[Bool]>()
    let banListQueue = Queue<[String]>()

    banIndexQueue.add(0)
    checkQueue.add([Bool] (repeating: false, count: user_id.count))
    banListQueue.add([])
    
    one: while !banIndexQueue.isEmpty() {
        let banIndex = banIndexQueue.poll()
        var check = checkQueue.poll()
        var banList = banListQueue.poll()
        
        if banIndex == banned_id.count {
            answerSet.insert(banList.sorted())
            continue one
        }
        
        let regex = "^" + banned_id[banIndex].replacingOccurrences(of: "*", with: ".") + "$"
        
        two: for i in 0..<user_id.count {
            if check[i] { continue }
            
            let currentUser = user_id[i]
            
            guard let _ = currentUser.range(of: regex, options: .regularExpression) else {
                continue two
            }
            
            banList.append(currentUser)
            check[i] = true
            
            banIndexQueue.add(banIndex + 1)
            checkQueue.add(check)
            banListQueue.add(banList)
            
            check[i] = false
            banList.removeLast()
            
        }
    }
    
    return answerSet.count
}

class Queue<T> {
    private var right: [T] = []
    private var left: [T] = []

    func add (_ item: T) {
        self.right.append(item)
    }

    func poll () -> T {
        if left.count == 0 {
            left = right.reversed()
            right.removeAll()
        }

        return left.removeLast()
    }

    func isEmpty () -> Bool {
        return left.count == 0 && right.count == 0
    }
}
